Sources/Common/Core/DataArray/index.js

```diff
@@ -313,44 +318,59 @@ function vtkDataArray(publicAPI, model) {
 // Object factory
 // ----------------------------------------------------------------------------
 
-const DEFAULT_VALUES = {
-  name: '',
-  numberOfComponents: 1,
-  size: 0,
-  dataType: DefaultDataType,
-  rangeTuple: [0, 0],
-  // values: null,
-  // ranges: null,
-};
+function defaultValues(initialValues) {
+  return {
+    name: '',
+    numberOfComponents: 1,
+    size: 0,
+    dataType: DefaultDataType,
+    rangeTuple: [0, 0],
+    // values: macro.newTypedArray(DefaultValues),
+    // ranges: null,
+    ...initialValues,
+  };
+}
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
-  Object.assign(model, DEFAULT_VALUES, initialValues);
-
-  if (!model.empty && !model.values && !model.size) {
+  if (!initialValues.empty && !initialValues.values && !initialValues.size) {
     throw new TypeError(
       'Cannot create vtkDataArray object without: size > 0, values'
     );
   }
-
-  if (!model.values) {
-    model.values = macro.newTypedArray(model.dataType, model.size);
-  } else if (Array.isArray(model.values)) {
-    model.values = macro.newTypedArrayFrom(model.dataType, model.values);
-  }
-
-  if (model.values) {
-    model.size = model.values.length;
-    model.dataType = getDataType(model.values);
-  }
+  Object.assign(initialValues, defaultValues(initialValues));
 
   // Object methods
   macro.obj(publicAPI, model);
   macro.set(publicAPI, model, ['name', 'numberOfComponents']);
 
+  model.dataType = initialValues.dataType
+    ? initialValues.dataType
+    : DefaultDataType;
+  if (!initialValues.values) {
+    initialValues.values = macro.newTypedArray(
+      model.dataType,
+      initialValues.size
+    );
+  } else if (Array.isArray(initialValues.values)) {
+    initialValues.values = macro.newTypedArrayFrom(
+      model.dataType,
+      initialValues.values
+    );
+  }
+
   // Object specific methods
   vtkDataArray(publicAPI, model);
+  // We need to call setData manually here because it is not a setter for a property
+  // in itself but calling it is still the proper way to set the datas of the model
+  publicAPI.setData(initialValues.values, initialValues.numberOfComponents);
```

Comment:
@jourdain , to avoid this exception, wouldn't it make sense to expose a convenient `setValues()` setter that simply calls setData() ? e.g. `publicAPI.setValues = (v) => publicAPI.setData(v);`

Reply:
setData setup the tuple size. So not keen to change the model name to data either.

Reply:
I made an update so that the property `values` is now `data` in DataArray and `Points` and `CellArray` to keep an example of child classes refactoring running. That change needed the `initialValues` to be sorted so that setData is called lastly (to be able to set properly the other properties). The update on the other classes using DataArray will be needed to be done.
Some other classes work similarly as DataArray (with a property `values` and a function `setData`), a similar change thus will be needed to be done.

Reply:
Sorry that I wasn't clear earlier, but I do-not want to have `values` renamed to `data` especially because setData set 2 things rather than setting only the values.